contracts/modules/VersionManager.sol

```diff
@@ -0,0 +1,296 @@
+
+    // Last bundle version
+    uint256 public lastVersion;
+    // Current bundle version for a wallet
+    mapping(address => uint256) public walletVersions; // [wallet] => [version]
+    // Features per version
+    mapping(address => mapping(uint256 => bool)) public isFeatureInVersion; // [feature][version] => bool
+    // Features requiring initialization for a wallet
+    mapping(uint256 => address[]) public featuresToInit; // [version] => [features]
+
+    // Supported static call signatures
+    mapping(uint256 => bytes4[]) public staticCallSignatures; // [version] => [sigs]
+    // Features executing static calls
+    mapping(uint256 => mapping(bytes4 => address)) public staticCallExecutors; // [version][sig] => [feature]
+
+    // Authorised Storages
+    mapping(address => bool) public isStorage; // [storage] => bool
+
+    event VersionAdded(uint256 _version);
+    event WalletUpgraded(address _wallet, uint256 _version);
+
+    // The Module Registry
+    IModuleRegistry private registry;
+
+    /* ***************** Modifiers ************************* */
+
+    modifier onlyFeature(address _wallet) {
+        require(isFeatureAuthorised(_wallet, msg.sender), "VM: sender should be authorized feature");
+        _;
+    }
+
+    /* ***************** Constructor ************************* */
+
+    constructor(
+        IModuleRegistry _registry,
+        ILockStorage _lockStorage,
+        IGuardianStorage _guardianStorage,
+        ITransferStorage _transferStorage,
+        ILimitStorage _limitStorage
+    )
+        BaseFeature(_lockStorage, IVersionManager(address(this)), NAME)
+        public
+    {
+        registry = _registry;
+
+        // Add initial storages
+        if(address(_lockStorage) != address(0)) { 
+            addStorage(address(_lockStorage));
+        }
+        if(address(_guardianStorage) != address(0)) { 
+            addStorage(address(_guardianStorage));
+        }
+        if(address(_transferStorage) != address(0)) {
+            addStorage(address(_transferStorage));
+        }
+        if(address(_limitStorage) != address(0)) {
+            addStorage(address(_limitStorage));
+        }
+    }
+
+    /* ***************** onlyOwner ************************* */
+
+    /**
+     * @inheritdoc IFeature
+     */
+    function recoverToken(address _token) external override onlyOwner {
+        uint total = ERC20(_token).balanceOf(address(this));
+        _token.call(abi.encodeWithSelector(ERC20(_token).transfer.selector, msg.sender, total));
+    }
+
+    /**
+     * @notice Lets the owner add a new version, i.e. a new bundle of features
+     * @param _features the list of features included in the new version
+     * @param _featuresToInit the subset of features that need to be initialized for a wallet
+     */
+    function addVersion(address[] calldata _features, address[] calldata _featuresToInit) external onlyOwner {
```

Comment:
I am still a bit worried about this `_featuresToInit` parameter. I understand that it saves some gas on creation/upgrade (~2000 gas per feature) but it is error prone and could create/update wallets to an inconsistent state if we forget to set it correctly. 

Maybe we should update the deployment script to better automate the generation of the `_featuresToInit` array by e.g. adding features to the array while they are created.

Reply:
For the record, according to benchmarks, using `_featuresToInit` saves:
- 7344 gas per new feature during a wallet upgrade
- 6627 gas per feature during a wallet creation

Reply:
To be a bit less error-prone, we could change the deployment script to list the features that do NOT require init instead of the those the DO require init.

```
  // Add Features to Version Manager
  const newFeatures = [
    GuardianManagerWrapper.contractAddress,
    LockManagerWrapper.contractAddress,
    RecoveryManagerWrapper.contractAddress,
    ApprovedTransferWrapper.contractAddress,
    TransferManagerWrapper.contractAddress,
    TokenExchangerWrapper.contractAddress,
    NftTransferWrapper.contractAddress,
    CompoundManagerWrapper.contractAddress,
    MakerV2ManagerWrapper.contractAddress,
    RelayerManagerWrapper.contractAddress,
  ];
  const newFeaturesWithNoInit = [ // all features except the TransferManager
    GuardianManagerWrapper.contractAddress,
    LockManagerWrapper.contractAddress,
    RecoveryManagerWrapper.contractAddress,
    ApprovedTransferWrapper.contractAddress,
    TokenExchangerWrapper.contractAddress,
    NftTransferWrapper.contractAddress,
    CompoundManagerWrapper.contractAddress,
    MakerV2ManagerWrapper.contractAddress,
    RelayerManagerWrapper.contractAddress,
  ];
  const newFeatureToInit = newFeatures.filter((f) => !newFeaturesWithNoInit.includes(f));
  const VersionManagerAddVersionTx = await VersionManagerWrapper.contract.addVersion(newFeatures, newFeatureToInit, { gasPrice });
  await VersionManagerWrapper.verboseWaitForTransaction(VersionManagerAddVersionTx, "Adding New Version");

```

Reply:
I'm surprised by the cost of calling init (~7k). Does this include the `TransferManager` or is this number valid for `init` methods only emitting an event? I would have expected the cost to be ~3k (1 call + 1 event).

Which reminds me that clients where using the `FeatureInitialised` (or `ModuleInitialised`) event to detect the features of a wallet. We need to make sure that it is no longer the case and that they are now using the `VersionAdded` event instead.

Reply:
I had checked with the clients that `ModuleInitialised` was in fact not used.

Reply:
It's the cost per feature for features with an `init` method that only emits an event. Also surprised by the 6-7k cost to be honeest. Roughly speaking that should be the cost of 3 reads ("feature address" + "is feature already initialized" + "is caller the version manager" = 3*800) + 1 call (2000) + 1 event (535, which we can remove btw), so I would have expected more like 5k. The rest seems to be overhead due to the for loop.

Reply:
So we should be able to save at least 800+535=1335 gas by making the default init method totally empty.

Reply:
Gas savings provided by _featuresToInit, when the default `init` does nothing at all:

4635 gas per new feature during a wallet upgrade
5464 gas per feature during a wallet creation

Reply:
Let's leave it as it is then. It seems that it is worth the gas saving.